x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/create_security_rule_type_factory.ts

```diff
@@ -0,0 +1,325 @@
+
+      const esClient = scopedClusterClient.asCurrentUser;
+
+      const ruleStatusClient = ruleStatusSavedObjectsClientFactory(savedObjectsClient);
+      const ruleStatusService = await ruleStatusServiceFactory({
+        alertId,
+        ruleStatusClient,
+      });
+
+      const ruleSO = await savedObjectsClient.get('alert', alertId);
+
+      const {
+        actions,
+        name,
+        schedule: { interval },
+      } = ruleSO.attributes;
+      const refresh = actions.length ? 'wait_for' : false;
+
+      const buildRuleMessage = buildRuleMessageFactory({
+        id: alertId,
+        ruleId,
+        name,
+        index: indexAlias,
+      });
+
+      logger.debug(buildRuleMessage('[+] Starting Signal Rule execution'));
+      logger.debug(buildRuleMessage(`interval: ${interval}`));
+
+      let wroteWarningStatus = false;
+      await ruleStatusService.goingToRun();
+
+      let result = createResultObject(state);
+
+      // check if rule has permissions to access given index pattern
+      // move this collection of lines into a function in utils
+      // so that we can use it in create rules route, bulk, etc.
+      try {
+        if (!isMachineLearningParams(params)) {
+          const index = params.index;
+          const hasTimestampOverride = !!timestampOverride;
+
+          // TODO: Input Index
+          const inputIndices = ['test'];
+
+          const [privileges, timestampFieldCaps] = await Promise.all([
+            checkPrivilegesFromEsClient(esClient, inputIndices),
+            esClient.fieldCaps({
+              index: index ?? ['*'],
+              fields: hasTimestampOverride
+                ? ['@timestamp', timestampOverride as string]
+                : ['@timestamp'],
+              include_unmapped: true,
+            }),
+          ]);
+
+          fold<Error, Promise<boolean>, void>(
+            async (error: Error) => logger.error(buildRuleMessage(error.message)),
+            async (status: Promise<boolean>) => (wroteWarningStatus = await status)
+          )(
+            flow(
+              () =>
+                tryCatch(
+                  () =>
+                    hasReadIndexPrivileges(privileges, logger, buildRuleMessage, ruleStatusService),
+                  toError
+                ),
+              chain((wroteStatus: unknown) =>
+                tryCatch(
+                  () =>
+                    hasTimestampFields(
+                      wroteStatus as boolean,
+                      hasTimestampOverride ? (timestampOverride as string) : '@timestamp',
+                      name,
+                      timestampFieldCaps,
+                      inputIndices,
+                      ruleStatusService,
+                      logger,
+                      buildRuleMessage
+                    ),
+                  toError
+                )
+              )
+            )() as Either<Error, Promise<boolean>>
```

Comment:
Is it possible to rewrite this logic without using `fp-ts`? It's quite hard to understand what is going on here. I think plain `try { ... } catch { ... }` would be much more readable.

Reply:
@xcrzx As per @peluja1012 's recommendation, I think we'll leave this for now. It was pre-existing code that I duplicated here. We may want to tackle it later, but it's not super high priority right now.

Reply:
@xcrzx haha I [wrote this] actually(https://github.com/elastic/kibana/pull/87489#discussion_r560316097)... yeah I think writing this in an imperative manner would be easier to read. I was in the process of cleaning some of this up and wanted to use fp-ts to drive the re-write but then rac happened and this was left as-is. Agreed that imperatively this reads a lot better.